File: core/src/main/java/javabot/javadoc/JavadocApi.java
package javabot.javadoc;

import javabot.model.Persistent;
import org.bson.types.ObjectId;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import org.mongodb.morphia.annotations.Index;
import org.mongodb.morphia.annotations.Indexes;
import org.mongodb.morphia.annotations.PrePersist;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.List;

@Entity(value = "apis", noClassnameStored = true)
@Indexes({
    @Index(value = "name", unique = true),
    @Index(value = "upperName", unique = true)
})
public class JavadocApi implements Persistent {
  private static final Logger log = LoggerFactory.getLogger(JavadocApi.class);

  @Id
  private ObjectId id;

  private String name;

  private String upperName;

  private String baseUrl;

  private String downloadUrl;

  public static final List<String> JDK_JARS = Arrays.asList("rt.jar", "jce.jar");

  public JavadocApi() {
  }

  public JavadocApi(final String apiName, final String url, final String downloadUrl) {
    name = apiName;
    baseUrl = url.endsWith("/") ? url : url + "/";
    this.downloadUrl = downloadUrl;
  }

  public ObjectId getId() {
    return id;
  }

  public void setId(final ObjectId id) {
    this.id = id;
  }

  public String getName() {
    return name;
  }

  public void setName(final String name) {
    this.name = name;
  }

  public String getBaseUrl() {
    return baseUrl;
  }

  public void setBaseUrl(final String baseUrl) {
    this.baseUrl = baseUrl;
  }

  public String getDownloadUrl() {
    return downloadUrl;
  }

  public void setDownloadUrl(String downloadUrl) {
    this.downloadUrl = downloadUrl;
  }

  @PrePersist
  public void uppers() {
    upperName = name.toUpperCase();
  }

  @Override
  public String toString() {
    return name;
  }
}

File: src/main/kotlin/javabot/javadoc/JavadocApi.kt
package javabot.javadoc

import javabot.model.ApiEvent
import javabot.model.ApiEvent.Companion
import javabot.model.Persistent
import org.bson.types.ObjectId
import org.mongodb.morphia.annotations.Entity
import org.mongodb.morphia.annotations.Field
import org.mongodb.morphia.annotations.Id
import org.mongodb.morphia.annotations.Index
import org.mongodb.morphia.annotations.IndexOptions
import org.mongodb.morphia.annotations.Indexes
import org.mongodb.morphia.annotations.PrePersist

@Entity(value = "apis", noClassnameStored = true)
@Indexes(Index(fields = arrayOf(Field("name")), options = IndexOptions(unique = true)),
      Index(fields = arrayOf(Field("upperName")), options = IndexOptions(unique = true)))
class JavadocApi : Persistent {

    @Id
    var id: ObjectId = ObjectId()

    lateinit var name: String

    private var upperName: String? = null

    lateinit var baseUrl: String

    lateinit var groupId: String

    lateinit var artifactId: String

    lateinit var version: String

    private constructor() {
        groupId = ""
        artifactId = ""
        version = ""
    }

    constructor(apiName: String, url: String, groupId: String, artifactId: String, version: String) {
        name = apiName
        this.groupId = groupId
        this.artifactId = artifactId
        this.version = if (name == "JDK") System.getProperty("java.specification.version") else version
        baseUrl = if (url.endsWith("/")) url else url + "/"
        baseUrl = "${baseUrl}javadoc/$apiName/${this.version}/"
    }

    @PrePersist fun uppers() {
        upperName = name.toUpperCase()
    }

    override fun toString(): String {
        return name
    }
}
